sphinx/tutorial/rose/furthertopics/failif-warnif.rst

```diff
@@ -0,0 +1,240 @@
+of ``total_weight_kg`` and ``fuelless_weight_kg``.
+
+Re-run the check by clicking
+:menuselection:`Metadata --> Check fail-if, warn-if`. An error
+dialog will appear, and the ``total_weight_kg`` setting will have
+an error flag.
+
+However, neither of these are very informative, other than quoting
+the metadata.
+
+Change the ``fail-if`` line to:
+
+.. code-block:: rose
+
+   fail-if=this < namelist:rocket=fuelless_weight_kg * 2.7183**(env=ORBITAL_SPEED_MS / (9.8 * namelist:rocket=specific_impulse_s));  # Fuel mass ratio or specific impulse too low to achieve orbit.
+
+If you reload the metadata and run the check again, the error
+messages will include the helpful text.
+
+You can also check the ``fail-if`` metadata by running
+``rose macro --validate`` or ``rose macro -V`` in a terminal,
+inside the app directory. Try saving the configuration in a failed
+state, and then run the command.
+
+
+``warn-if``
+-----------
+
+The ``warn-if`` metadata setting is exactly the same as ``fail-if``,
+but gives out warnings instead of errors - from a user perspective,
+this means something like "Watch out" rather than "This is wrong".
+
+Let's try adding something for ``namelist:rocket=battery_levels``.
+
+Open the metadata file ``meta/rose-meta.conf`` in a text editor,
+and add this line to the ``[namelist:rocket=battery_levels]`` section:
+
+.. code-block:: rose
+
+   warn-if=namelist:rocket=battery_levels(1) < 75 or namelist:rocket=battery_levels(2) < 75;
+
+This uses a special syntax for referencing the individual array
+elements in ``battery_levels``.
+
+If the first array element value and/or the second array element
+value of ``battery_levels`` is less than 75% full, a warning will
+be produced when the check is run.
+
+We already know the shorthand syntax this, so rephrase the metadata to:
+
+.. code-block:: rose
+
+   warn-if=this(1) < 75 or this(2) < 75;
+
+Save the metadata file and then reload the config editor metadata.
+Click :menuselection:`Metadata --> Check fail-if, warn-if` - a
+warning dialog should appear, and there should be a warning flag
+for the ``battery_levels`` option.
+
+For large arrays, it can sometimes be convenient to use whole-array
+operations - the ``fail-if`` and ``warn-if`` syntax includes ``any()``
+and ``all()``.
+
+We can change the ``warn-if`` setting to:
+
+.. code-block:: rose
+
+   warn-if=any(this < 75);
+
+which will flag a warning if any ``battery_levels`` array element
+values are less than 75.
+
+
+Multiple Expressions
+--------------------
+
+In both ``fail-if`` and ``warn-if``, expressions can be chained
+using the Python operator or, or you can separate them to give
+clearer error/warning messages. Using our ``battery_levels`` example
+again, change the setting to:
+
+.. code-block:: rose
+
+   warn-if=any(this < 75);
+          =all(this > 95);
+
+This gives a warning if any elements are less than 75, and a
+separate warning if all elements are greater than 95 (we don't
```

Comment:
Change 'gives' to 'will give' or 'would give' to avoid potential confusion as obviously with the default (80,60) variables only the former warning would show up if the reader tested this on the suite as it is.